### PR TITLE
Add support for app-factory usage and test it.

### PR DESCRIPTION
**Pull request by tonyseek**
@kennethreitz Please review. Thanks.

### CODE REVIEW

**Comment by kennethreitz** [comment]
This looks like a nice set of changes. Will have to set some time aside to review.

File: flask_sslify.py [hunk 1]
```diff
@@ -2,29 +2,43 @@
...
+    def permanent(self):
+        return self.app.config['SSLIFY_PERMANENT']
+
+    @property
+    def skip_list(self):
+        return self.app.config['SSLIFY_SKIPS']
+
     def init_app(self, app):
         """Configures the configured Flask app to enforce SSL."""
+        app.config.setdefault('SSLIFY_SUBDOMAINS', self.defaults['subdomains'])
```
Comment (joshfriend) [inline flask_sslify.py]: @kennethreitz This PR will fix the regression introduced by #28 ([explanation](https://github.com/kennethreitz/flask-sslify/pull/28/files#r53332282))

**Comment by ryan-lane** [comment]
This PR looks good to me. +1. Thank you for adding tests!

**Comment by kennethreitz** [comment]
would you mind removing `tox.ini` and `setup.cfg`? I will be happy to merge then :)

**Comment by tonyseek** [comment]
@kennethreitz I'm sorry for my late response.

I have removed the `tox.ini` and `setup.py`. Could you review it? Thanks!

**Review by richmondwang — APPROVED** [review]
this fixes the app factory pattern

**Comment by kennethreitz** [comment]
merge conflict

**Comment by tonyseek** [comment]
conflict resolved in a rebasing

File: flask_sslify.py [hunk 2]
```diff
@@ -2,32 +2,48 @@
...
     def init_app(self, app):
         """Configures the specified Flask app to enforce SSL."""
-        app.config.setdefault('SSLIFY_SUBDOMAINS', False)
-        app.config.setdefault('SSLIFY_PERMANENT', False)
-        app.config.setdefault('SSLIFY_SKIPS', None)
-
-        self.hsts_include_subdomains = self.hsts_include_subdomains or app.config['SSLIFY_SUBDOMAINS']
-        self.permanent = self.permanent or self.app.config['SSLIFY_PERMANENT']
-        self.skip_list = self.skip_list or self.app.config['SSLIFY_SKIPS']
+        app.config.setdefault('SSLIFY_AGE', self.defaults['AGE'])
```
Comment (joshfriend) [inline flask_sslify.py]: None of the tests are running, this line is raising `KeyError: 'AGE'`. It should be `age` lowercase.

File: tests/test_factory.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+def test_config(app_factory):
+    assert app_factory.sslify.hsts_include_subdomains is False
+    assert app_factory.sslify.permanent is True
+    assert app_factory.sslify.skip_list is None
+
+
+def test_redirection(app_factory):
+    client = app_factory.app.test_client()
+    r = client.get('/')
+    assert r.status_code == 301
```
Comment (joshfriend) [inline tests/test_factory.py:54]: The redirect tests do not pass, the app is returning 200

File: tests/test_factory.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+from flask import Flask
+from flask_sslify import SSLify
+from pytest import yield_fixture
+
+
+class TestingContext(object):
```
Comment (joshfriend) [inline tests/test_factory.py]: I'd suggest calling this class `FixtureContext` or something else that doesn't begin with `Test`. py.test is trying to collect tests from it and is outputting a warning

**Review by joshfriend — CHANGES_REQUESTED** [review]

**Comment by tonyseek** [comment]
@joshfriend Sorry for that. I forgot to re-run testing before pushing. I fixed them all.

**Review by joshfriend — APPROVED** [review]

**Comment by rconradharris** [comment]
This fix looks good, would be great to get this into PyPI at some point.

**Pull request merged**